File: solutions/33. Search in Rotated Sorted Array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
    int n = nums.size();
    if(n == 0){
        return -1;
    }
    int low = 0, high = n-1;
    int firstel = nums[0];
    while(low <= high)
    {
        int mid = low + (high-low)/2;
        int value = nums[mid];
        if(target == value){
            return mid;
        }
        int am_big = value >= firstel;
        int target_bg = target >= firstel;
        if(am_big == target_bg) {
            if(value < target) {
                low = mid + 1;
            }else {
                high = mid - 1;
            }
        }else {
            if(am_big) {
                low = mid + 1;
            }else {
                high = mid - 1;
            }
        }
    }
        return -1;
    }
};

File: solutions/33. Search in Rotated Sorted Array.java
class Solution {
	public int search(int[] nums, int target) {
		int lo = 0;
		int hi = nums.length - 1;

		while (lo < hi) {
			int mid = (lo + hi) / 2;
			int midelem = nums[mid];

			if (midelem == target)
				return mid;

			if (nums[lo] <= midelem) {
				if (nums[lo] <= target && target < midelem) {
					hi = mid - 1;
				} else {
					lo = mid + 1;
				}
			} else {
				if (midelem < target && target <= nums[hi]) {
					lo = mid + 1;
				} else {
					hi = mid - 1;
				}
			}
		}

		return nums[lo] == target ? lo : -1;
	}
}
